File: api/src/main/java/net/blueshell/api/validation/file/FileSize.java
package net.blueshell.api.validation.file;

import jakarta.validation.Constraint;
import jakarta.validation.Payload;

import java.lang.annotation.*;

@Documented
@Target({ElementType.FIELD, ElementType.PARAMETER})
@Retention(RetentionPolicy.RUNTIME)
@Constraint(validatedBy = {FileSizeValidator.class})
public @interface FileSize {

    String message() default "File size must be between {min} and {max} bytes";

    Class<?>[] groups() default {};

    Class<? extends Payload>[] payload() default {};

    /**
     * Inclusive min (bytes)
     */
    long min() default 0L;

    /**
     * Inclusive max (bytes)
     */
    long max();

    /**
     * Whether an empty upload (isEmpty) is allowed
     */
    boolean allowEmpty() default false;
}

File: api/src/main/kotlin/net/blueshell/api/feature/file/validation/FileSize.kt
package net.blueshell.api.feature.file.validation

import jakarta.validation.Constraint
import jakarta.validation.Payload
import kotlin.reflect.KClass

@MustBeDocumented
@Target(AnnotationTarget.FIELD, AnnotationTarget.VALUE_PARAMETER)
@Retention(AnnotationRetention.RUNTIME)
@Constraint(validatedBy = [FileSizeValidator::class])
annotation class FileSize(
    val message: String = "File size must be between {min} and {max} bytes",
    val groups: Array<KClass<*>> = [],
    val payload: Array<KClass<out Payload>> = [],
    /**
     * Inclusive min (bytes)
     */
    val min: Long = 0L,
    /**
     * Inclusive max (bytes)
     */
    val max: Long,
    /**
     * Whether an empty upload (isEmpty) is allowed
     */
    val allowEmpty: Boolean = false
)

